### ISSUE TITLE
Inconsistent while block due to grammar? in python3

### ISSUE THREAD

**Issue by x0ret** [rocky/python-uncompyle6#228]
## Description

Suppose following src:
```python
def func():
    while not x:
        print(x)
```

`uncompyle6`  decompiles this to :

```
def func():
    while 1:
        if not x:
            print(x)
```

I think we need a better representation or at least a `break` statement.

based on my limited debugging experience here, Please consider this module disassembly:

```
   3       0  SETUP_LOOP           18  'to 18'
           2  LOAD_GLOBAL              'x'
           4  POP_JUMP_IF_TRUE     16  'to 16'

   4       6  LOAD_GLOBAL              'print'
           8  LOAD_GLOBAL              'x'
          10  CALL_FUNCTION         1  ''
          12  POP_TOP
          14  JUMP_BACK             2  'to 2'
          16  POP_BLOCK
        18_0  COME_FROM_LOOP        0  '0'
          18  LOAD_CONST               None
          20  RETURN_VALUE
```

And we have following grammer in `parse3.py` at [L#154](https://github.com/rocky/python-uncompyle6/blob/master/uncompyle6/parsers/parse3.py#L154):

```
        iflaststmt ::= testexpr c_stmts_opt JUMP_ABSOLUTE

        iflaststmtl ::= testexpr c_stmts_opt JUMP_BACK [1]
        iflaststmtl ::= testexpr c_stmts_opt JUMP_BACK COME_FROM_LOOP
        iflaststmtl ::= testexpr c_stmts_opt JUMP_BACK POP_BLOCK [2]
```

based on disassembly if we omit [1], [2] (commenting them out) we have following code:

```
def func():
    while not x:
        print(x)
```

I'm not sure what is the best fix here. I also checked 3.7 version result is the same.

for reference we have following grammer before fix(?):

```
         2-12  iflaststmtl ::= testexpr c_stmts_opt (7)
L.  3:   6     _ifstmts_jump ::= c_stmts_opt (7)
         2     lastl_stmt ::= iflaststmtl (7)
         2-12  ifstmt ::= testexpr _ifstmts_jump (7)
         2     l_stmts ::= lastl_stmt (7)
         2     stmt ::= ifstmt (7)
         2     _stmts ::= stmt (7)
         2-14  iflaststmtl ::= testexpr c_stmts_opt JUMP_BACK (8)
         2     lastl_stmt ::= iflaststmtl (8)
         2     l_stmts ::= lastl_stmt (8)
         2     l_stmts_opt ::= l_stmts (8)
         2-16  iflaststmtl ::= testexpr c_stmts_opt JUMP_BACK POP_BLOCK (9)
         2     lastl_stmt ::= iflaststmtl (9)
         2     l_stmts ::= lastl_stmt (9)
         2     l_stmts_opt ::= l_stmts (9)
               whilestmt ::= SETUP_LOOP testexpr l_stmts_opt JUMP_BACK POP_BLOCK COME_FROM_LOOP (10)
               whileTruestmt ::= SETUP_LOOP l_stmts_opt JUMP_BACK POP_BLOCK COME_FROM_LOOP (10)
               while1stmt ::= SETUP_LOOP l_stmts POP_BLOCK COME_FROM_LOOP (10)
               while1stmt ::= SETUP_LOOP l_stmts COME_FROM_LOOP (10)
```

after patch:

```
         2-12  iflaststmtl ::= testexpr c_stmts_opt (7)
L.  3:   6     _ifstmts_jump ::= c_stmts_opt (7)
         2     lastl_stmt ::= iflaststmtl (7)
         2-12  ifstmt ::= testexpr _ifstmts_jump (7)
         2     l_stmts ::= lastl_stmt (7)
         2     stmt ::= ifstmt (7)
         2     _stmts ::= stmt (7)
               whilestmt ::= SETUP_LOOP testexpr l_stmts_opt JUMP_BACK POP_BLOCK COME_FROM_LOOP (10)
               whileTruestmt ::= SETUP_LOOP l_stmts_opt JUMP_BACK POP_BLOCK COME_FROM_LOOP (10)
```

what's your opinion @rocky ?

## How to Reproduce

Please use following commands:

```bash
$ cat test1.py
def func():
    while not x:
        print(x)
$ python -m compileall . ; uncompyle6 __pycache__/test1.cpython-36.pyc
Listing '.'...
# uncompyle6 version 3.3.2
# Python bytecode 3.6 (3379)
# Decompiled from: Python 3.6.7 (default, Oct 22 2018, 11:32:17)
# [GCC 8.2.0]
# Embedded file name: ./test1.py
# Compiled at: 2019-05-09 17:06:07
# Size of source mod 2**32: 46 bytes


def func():
    while 1:
        if not x:
            print(x)
# okay decompiling __pycache__/test1.cpython-36.pyc
```
## Expected behavior

We expect the condition to be in `while cond` form or a `while 1` with a break in the `if cond`.

## Environment

- Uncompyle6 version: master HEAD `b57ca392`
- Python version: `3.6.7`
- OS and Version: Ubuntu bionic

Thank you

**Comment by rocky** [comment]
Translating:

```
while not x:
     print(x)
```

as: 
bb
```

while 1:
     if not x:
          print(x)
```

is technically not wrong. It just is not the way many people would write the code. 

The problem with changing the grammar is that fixing this here may break something else. 

There are two better ways to address. The one I prefer is to fix this up as a phase in rewriting the syntax tree. If  it notices a "whileTrue with a single "ifnot" it could revise the tree to be "whilenot'. 

"if x: raise AssertError"  would also be turned into "assert"  And there are some existing transformations in pysource that would be cleaned up. 

The other way to handle this better might be in doing better control flow detection using the `python-control-flow` project. 

So unless there is another example which is demonstrably _wrong_, I'd prefer to leave this as is.

**Comment by rocky** [comment]
For what it's worth this has been fixed in decompile3 by adding a proper `COME_FROM`  from the `not x` test. That this isn't added is due to some earler 3.x versions I think erroneously requiring their not be the `COME_FROM`  for grammar structure detection.

So possibly this may be cleaned up in decompile3 and then backported here.

**Comment by rocky** [comment]
With commit f2501120 which starts to rermove crud, this also works on 3.8.0.

**Comment by rocky** [comment]
This was fixed sometime before fixed in 8eb4d6a576ce440ff3054f446603135d9e56a0c2   3.7.4 will go out soon.
